apps/aecore/src/aec_conductor.erl

```diff
@@ -0,0 +1,601 @@
+%%% Internal functions
+%%%===================================================================
+
+%%%===================================================================
+%%% Handle init options
+
+set_option(autostart, Options, State) ->
+    case get_option(autostart, Options) of
+        undefined   -> State;
+        {ok, true}  -> State#state{mining_state = running};
+        {ok, false} -> State#state{mining_state = stopped}
+    end;
+set_option(fetch_new_txs_from_pool, Options, State) ->
+    case get_option(fetch_new_txs_from_pool, Options) of
+        undefined -> State;
+        {ok, Val} -> State#state{fetch_new_txs_from_pool = Val}
+    end.
+
+get_option(Opt, Options) ->
+    case proplists:lookup(Opt, Options) of
+        none -> application:get_env(aecore, Opt);
+        {_, Val} -> {ok, Val}
+    end.
+
+%%%===================================================================
+%%% Worker handling
+%%% @private
+%%% @doc
+%%%
+%%% Worker functions are funs of arity 0 with a tag to determine the
+%%% type of worker. Tags are enforced to be 'singleton' (only one worker
+%%% allowed) or 'concurrent' (allow concurrent processes).
+%%%
+%%% The worker processes are monitored, and provides return values
+%%% through message passing. Return values are passed as messages, and
+%%% the reply is handled based on the tag. The worker fun does not
+%%% need to handle the message passing itself. This is taken care of
+%%% by the dispatcher.
+%%%
+%%% Note that when the reply is handled, the state is the current
+%%% server state, not the state in which the worker was
+%%% dispatched. Any consistency checks for staleness must be handled
+%%% in the reply handler.
+%%%
+%%% Workers can be killed (e.g., on preemption because of a changed
+%%% chain) based on tag. Note that since the worker might have sent an
+%%% answer before it is killed, it is good to check answers for
+%%% staleness. TODO: This could be done by the framework.
+
+dispatch_worker(Tag, Fun, State) ->
+    case is_worker_allowed(Tag, State) of
+        true ->
+            Wrapper = wrap_worker_fun(Fun),
+            {Pid, Ref} = spawn_monitor(Wrapper),
+            State1 = maybe_block_tag(Tag, State),
+            Workers = orddict:store(Pid, {Tag, Ref}, State1#state.workers),
+            State#state{workers = Workers};
+        false ->
+            epoch_mining:error("Disallowing dispatch of aditional ~p worker",
+                               [Tag]),
+            State
+    end.
+
+is_worker_allowed(Tag, State) ->
+    not ordsets:is_element(Tag, State#state.blocked_tags).
+
+maybe_block_tag(Tag, #state{blocked_tags = B} = State) ->
+    case tag_concurrency(Tag) of
+        concurrent -> State;
+        singleton  -> State#state{blocked_tags = ordsets:add_element(Tag, B)}
+    end.
+
+tag_concurrency(create_block_candidate) -> singleton;
+tag_concurrency(mining)                 -> singleton;
+tag_concurrency(post_block)             -> concurrent;
```

Comment:
How will concurrent post block be handled when the chain is turned into a functional ADT instead of a server?

Reply:
Good point. Now it works, but it will probably break down then. The validation could be done in a worker, but the insert should probably be atomic.